pkg/transformers/refvars_helper.go

```diff
@@ -0,0 +1,135 @@
+package transformers
+
+import (
+	"fmt"
+)
+
+// TODO(jeb): Investigate the usage of obj->json + jsonPatch
+// deepMerge merges two objects together
+func deepMerge(dstVal interface{}, srcVal interface{}) (interface{}, error) {
+	switch srcValType := srcVal.(type) {
+	case nil:
+		// preserving non nil type
+		return dstVal, nil
+	case []interface{}:
+		return castAndMergeSlice(dstVal, srcVal)
+	case map[string]interface{}:
+		return castAndMergeMap(dstVal, srcVal)
+	default:
+		switch dstValType := dstVal.(type) {
+		case nil:
+			return srcVal, nil
+		default:
+			if dstValType != srcValType {
+				// return dstVal,
+				//	fmt.Errorf("Conflicting type. Unable to merge key %T %T", srcValType, dstValType)
+				dstVal = srcVal
+			} else {
+				dstVal = srcVal
+			}
+		}
+	}
+
+	return dstVal, nil
+}
+
+func castAndMergeSlice(dstVal interface{}, srcVal interface{}) (interface{}, error) {
+	switch dstValType := dstVal.(type) {
+	case []interface{}:
+		dstSlice := dstVal.([]interface{})
+		srcSlice := srcVal.([]interface{})
+		mergedSlice, err := deepMergeSlice(dstSlice, srcSlice)
+		if err != nil {
+			return dstVal, err
+		}
+		return mergedSlice, nil
+	case nil:
+		return srcVal, nil
+	default:
+		return dstVal, fmt.Errorf("Conflicting type. Unable to merge %T", dstValType)
+	}
+}
+
+func deepMergeSlice(dstSlice []interface{}, srcSlice []interface{}) ([]interface{}, error) {
+	if len(dstSlice) != len(srcSlice) {
+		return dstSlice,
+			fmt.Errorf("Conflicting arrays. Unable to merge array with conflicting length")
+	}
+	for id, dstVal := range srcSlice {
+		mergedVal, err := deepMerge(dstVal, srcSlice[id])
+		if err != nil {
+			return dstSlice, err
+		}
+		dstSlice[id] = mergedVal
+	}
+
+	return dstSlice, nil
+}
+
+func castAndMergeMap(dstVal interface{}, srcVal interface{}) (interface{}, error) {
+	switch dstValType := dstVal.(type) {
+	case map[string]interface{}:
+		dstMap := dstVal.(map[string]interface{})
+		srcMap := srcVal.(map[string]interface{})
+		mergedMap, err := deepMergeMap(dstMap, srcMap)
+		if err != nil {
+			return dstVal, err
+		}
+		return mergedMap, nil
+	case nil:
+		return dstVal, nil
+	default:
+		return dstVal,
+			fmt.Errorf("Conflicting type. Unable to merge %T", dstValType)
+
+	}
+}
+
+func deepMergeMap(dstMap map[string]interface{}, srcMap map[string]interface{}) (map[string]interface{}, error) {
+	for key, srcVal := range srcMap {
+		if dstVal, ok := dstMap[key]; ok {
+			mergedVal, err := deepMerge(dstVal, srcVal)
+			if err != nil {
+				return dstMap, err
+			}
+			dstMap[key] = mergedVal
+		} else {
+			dstMap[key] = srcVal
+		}
+	}
+	return dstMap, nil
+}
+
+// TODO(jeb): Investigate the usage of obj->json->obj conversion
```

Comment:
@monopole Do you have any opinion on using a json patch strategy here? Maybe something like what's being done in the [k8sdeps/transformer/patch package](https://github.com/kubernetes-sigs/kustomize/blob/master/k8sdeps/transformer/patch/transformer.go#L66). I think we could do away with a lot of the code in this file if we switched to something similar.